Use `Write.TxOut` in `ErrBalanceTxInputResolutionConflicts`.

jonathanknowles opened the pull request:
## Issue

ADP-3184

## Description

This PR adjusts `ErrBalanceTxInputResolutionConflicts` to use `Write.TxOut` instead of the wallet primitive `TxOut` type.

jonathanknowles committed:
Add `ErrBalanceTxInRecentEra` type.

Use this type to hide the existential type variable of `ErrBalanceTx`.

jonathanknowles committed:
Add `IsRecentEra` constraint to `ErrBalanceTxInRecentEra`.

jonathanknowles committed:
Use `exceptionBalanceTx` to factor out common sub-expression.

jonathanknowles committed:
Test `prop_calculateFeePercentiles` for all recent eras.

jonathanknowles committed:
Test `feeEstimationRegressionSpec` for all recent eras.

jonathanknowles committed:
Add `IsRecentEra` constraint to `{Eq,Show}` instances for `ErrBalanceTx`.

jonathanknowles committed:
Use consistent formatting in `guardWalletUTxOConsistencyWith`.

jonathanknowles committed:
Use clearer names for outputs in `guardWalletUTxOConsistencyWith`.

jonathanknowles committed:
Use `Write.TxOut` in `ErrBalanceTxInputResolutionConflicts`.

jonathanknowles committed:
Remove `ErrBalanceTxInRecentEra` wrapper type.

Anviking commented on lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -991,8 +991,18 @@ decodeSealedTxSpec = describe "SealedTx serialisation/deserialisation" $ do
+        RecentEraBabbage ->
+            feeEstimationRegressionSpecInner @Cardano.BabbageEra era
+
+feeEstimationRegressionSpecInner
```
It seems unnecessary to define `feeEstimationRegressionSpecInner` separately from `feeEstimationRegressionSpec` to me. I'll create another PR

Anviking replied:
https://github.com/cardano-foundation/cardano-wallet/pull/4207

jonathanknowles replied:
@Anviking Thanks for the suggestion! I've cherry-picked your suggested commit into this branch. 👍🏻

Anviking commented on lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/Balance.hs:
```diff
@@ -902,14 +904,16 @@ balanceTransactionWithSelectionStrategyAndNoZeroAdaAdjustment
     guardWalletUTxOConsistencyWith u' = do
         let W.UTxO u = toWalletUTxO (recentEra @era) $ fromCardanoApiUTxO u'
-        let conflicts = lefts $ flip map (Map.toList u) $ \(i, o) ->
+        let conflicts = lefts $ flip map (Map.toList u) $ \(i, out1) ->
```
Changing `o` to `out` breaks the visual similarity with `i`. Is this really worth it?

jonathanknowles replied:
Here I was aiming for visual similarity between `out1` and `out2`:
```hs
                   Just out2 -> unless (out1 == out2) $ Left
                        ( toLedgerTxOut era out1
                        , toLedgerTxOut era out2
                        )
```

I can rename them to `o1` and `o2` if you prefer?

Anviking replied:
I'd prefer sticking to `o`, `o'`, or possibly `o1` or `o2` if you want them to have the same length

Anviking replied:
Although the added descriptiveness might be worth it as you have it 🤔 => 🤷‍♂️ approved

jonathanknowles replied:
Okay! Fixed in 6e3879e6acb184eaa40b20a057fbeae60266d760.

Anviking commented on lib/wallet/test/unit/Cardano/WalletSpec.hs:
```diff
@@ -759,11 +760,16 @@ walletListsOnlyRelatedAssets txId txMeta =
-prop_calculateFeePercentiles :: NonEmptyList (Maybe Coin) -> Property
-prop_calculateFeePercentiles (NonEmpty coins) =
+prop_calculateFeePercentiles
+    :: Write.AnyRecentEra
```
👍

Anviking commented on lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -1002,7 +1013,7 @@ feeEstimationRegressionSpec = describe "Regression tests" $ do
                     , shortfall = Ledger.Coin 100_000
                     }
-        result <- runExceptT (calculateFeePercentiles estimateFee)
+        result <- runExceptT (calculateFeePercentiles @_ @era estimateFee)
```
Not specific to this PR but just thinking about our direction: Maybe the need for additional type-application is a bad sign 🤔

Anviking committed:
Inline `feeEstimationRegressionSpecInner`

Anviking reviewed (commented):
lgtm modulo two comments

jonathanknowles committed:
Shorten names of outputs in `guardWalletUTxOConsistencyWith`.

In response to review feedback:

https://github.com/cardano-foundation/cardano-wallet/pull/4206/files#r1384613103
